riptide-metrics/src/main/java/org/zalando/riptide/metrics/MetricsPlugin.java

```diff
@@ -10,33 +12,63 @@
 import org.zalando.riptide.RequestArguments;
 import org.zalando.riptide.RequestExecution;
 
-import static java.util.Objects.nonNull;
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.collect.Iterables.concat;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
 import static org.apiguardian.api.API.Status.EXPERIMENTAL;
 import static org.zalando.fauxpas.FauxPas.throwingBiConsumer;
 
 @API(status = EXPERIMENTAL)
 public final class MetricsPlugin implements Plugin {
 
     private final MeterRegistry registry;
-    private final MetricsNameGenerator generator;
+    private final String metricName;
+    private final ImmutableList<Tag> defaultTags;
+    private final Clock clock;
+
+    public MetricsPlugin(final MeterRegistry registry) {
+        this(registry, "http.client.requests", ImmutableList.of());
+    }
 
-    public MetricsPlugin(final MeterRegistry registry, final MetricsNameGenerator generator) {
+    private MetricsPlugin(final MeterRegistry registry, final String metricName, final ImmutableList<Tag> defaultTags) {
         this.registry = registry;
-        this.generator = generator;
+        this.metricName = metricName;
+        this.defaultTags = defaultTags;
+        this.clock = registry.config().clock();
+    }
+
+    public MetricsPlugin withMetricName(final String metricName) {
+        return new MetricsPlugin(registry, metricName, defaultTags);
+    }
+
+    public MetricsPlugin withDefaultTags(final Tag... defaultTags) {
+        return withDefaultTags(ImmutableList.copyOf(defaultTags));
+    }
+
+    public MetricsPlugin withDefaultTags(final Iterable<Tag> defaultTags) {
+        return new MetricsPlugin(registry, metricName, ImmutableList.copyOf(defaultTags));
     }
 
     @Override
     public RequestExecution interceptBeforeRouting(final RequestArguments arguments, final RequestExecution execution) {
         return () -> {
             final Measurement measurement = new Measurement(arguments);
 
-            return execution.execute()
-                    .whenComplete(throwingBiConsumer((response, e) -> {
-                        if (nonNull(response)) {
-                            measurement.record(response);
-                        }
-                    }));
+            final CompletableFuture<ClientHttpResponse> future;
+
+            try {
+                future = execution.execute();
+            } catch (final IOException e) {
+                measurement.record(null, e);
```

Comment:
How do I test this?

Reply:
super specific unit test with super specific mocks... :deciduous_tree: